libraries/CryptoInterface/src/CryptoInterface.cpp

```diff
@@ -0,0 +1,566 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "CryptoInterface.h"
+#include <TypeConversion.h>
+
+#include <bearssl/bearssl.h>
+#include <assert.h>
+
+namespace TypeCast = esp8266::TypeConversion;
+
+namespace
+{
+  size_t _ctMinDataLength = 0;
+  size_t _ctMaxDataLength = 1024;
+
+  bool _warningsEnabled = true;
+  
+  br_hkdf_context _storedHkdfContext;
+  bool _hkdfContextStored = false;
+
+  uint8_t *defaultNonceGenerator(uint8_t *nonceArray, const size_t nonceLength)
+  {
+    /**
+     * The ESP32 Technical Reference Manual v4.1 chapter 24 has the following to say about random number generation (no information found for ESP8266):
+     * 
+     * "When used correctly, every 32-bit value the system reads from the RNG_DATA_REG register of the random number generator is a true random number.
+     * These true random numbers are generated based on the noise in the Wi-Fi/BT RF system.
+     * When Wi-Fi and BT are disabled, the random number generator will give out pseudo-random numbers.
+     * 
+     * When Wi-Fi or BT is enabled, the random number generator is fed two bits of entropy every APB clock cycle (normally 80 MHz).
+     * Thus, for the maximum amount of entropy, it is advisable to read the random register at a maximum rate of 5 MHz.
+     * A data sample of 2 GB, read from the random number generator with Wi-Fi enabled and the random register read at 5 MHz,
+     * has been tested using the Dieharder Random Number Testsuite (version 3.31.1).
+     * The sample passed all tests."
+     * 
+     * Since ESP32 is the sequal to ESP8266 it is unlikely that the ESP8266 is able to generate random numbers more quickly than 5 MHz when run at a 80 MHz frequency.
+     * A maximum random number frequency of 0.5 MHz is used here to leave some margin for possibly inferior components in the ESP8266.
+     * It should be noted that the ESP8266 has no Bluetooth functionality, so turning the WiFi off is likely to cause RANDOM_REG32 to use pseudo-random numbers.
+     * 
+     * It is possible that yield() must be called on the ESP8266 to properly feed the hardware random number generator new bits, since there is only one processor core available. 
+     * However, no feeding requirements are mentioned in the ESP32 documentation, and using yield() could possibly cause extended delays during nonce generation.
+     * Thus only delayMicroseconds() is used below.
+     */ 
```

Comment:
That's something I did not know about the HW.  Assumed it was driven off of the usual Zener.  Do you think the BSSL port needs to include this?

Reply:
Include yields or delays you mean? Good question. I think I have too little data to answer conclusively. I simply couldn't find any source on the topic other than the ESP32 docs, which may still be applicable since the two MCUs are closely related. However, it is possible that the random number generation happens in a different way for the ESP8266 compared to the ESP32, so I wouldn't completely rule out a Zener diode being used.

So until we have more data I guess the answer depends on the cost of making the changes. If they are too high, perhaps it is worthwhile seeking clarification from Espressif first.